rlclientlib/logger/http_transport_client.h

```diff
@@ -126,45 +154,60 @@ pplx::task<web::http::status_code> http_transport_client<TAuthorization>::http_r
   const auto stream = concurrency::streams::bytestream::open_istream(container);
   request.set_body(stream, container_size);
 
-  return _client->request(request).then([this, try_count](pplx::task<http_response> response) {
-    web::http::status_code code = status_codes::InternalError;
-    api_status status;
-
+  // lambda which examines the provided task and either 1) generates a replacement task to retry
+  // the request, or 2) passes the provided task downstream to emit its response
+  auto retry_request_on_failure_lambda = [this, try_count](pplx::task<http_response> response_task) -> pplx::task<http_response>
+  {
+    web::http::status_code response_code = status_codes::InternalError;
+    
     try
     {
-      code = response.get().status_code();
+      response_code = response_task.get().status_code();
     }
     catch (const std::exception& e)
     {
       TRACE_ERROR(_trace, e.what());
     }
 
-    // If the response is not the expected code then it has failed. Retry if possible otherwise report background
-    // error.
-    if (code != status_codes::Created && code != status_codes::NoContent)
+    bool success = (response_code >= status_codes::OK) && (response_code < status_codes::MultipleChoices);
+    if (success)
     {
-      // Stop condition of recurison.
-      if (try_count < _max_retries)
-      {
-        TRACE_ERROR(_trace, "HTTP request failed, retrying...");
+      return response_task;
+    }
 
-        // Yes, recursively send another request inside this one. If a subsequent request returns success we are
-        // good, otherwise the failure will propagate.
-        return send_request(try_count + 1).get();
-      }
-      else
-      {
-        auto msg = u::concat("(expected 201): Found ", code, ", failed after ", try_count, " retries.");
-        api_status::try_update(&status, error_code::http_bad_status_code, msg.c_str());
-        ERROR_CALLBACK(_error_callback, status);
+    // If the response is not success class then it has failed. Retry if possible otherwise report background error.
 
-        return code;
-      }
+    if (try_count >= _max_retries)
+    {
+      // We have exhausted retry attempts, log and return the task describing the failure
+      api_status status;
+      auto msg = u::concat("(expected 201): Found ", response_code, ", failed after ", try_count, " retries.");
+      api_status::try_update(&status, error_code::http_bad_status_code, msg.c_str());
+      ERROR_CALLBACK(_error_callback, status);
+
+      return response_task;
+    }
+
+    if (status_codes::TooManyRequests == response_code)
```

Comment:
do we care about 503?

Reply:
I filtered for 429 response trying to make the minimal change that would be effective for my use case.

There's dozens of things that could go wrong for any HTTP request, so if DNS is down, or network is down, or servers are down, probably doing an incremental backoff in retries would be sensible.

However, I'm more concerned against introducing large delays in this codepath, since there isn't a good cancellation or status route.  I'd prefer to avoid using large delays for all failure cases until cancel/status is handled better.

Reply:
yeah, I called out 503 since it is also related to overloaded server but it also depends on the specific server impl.

Reply:
@lalo I reworked the retry logic accept a configured maximum retry timespan, and now all failures retry until either the retry attempts or retry timespan is exhausted.

For my 429 case I'll ask for 99999 retries and give a large time budget instead.  For everyone else they can either keep the existing behavior or opt in to the new timeboxed retries.